oauthproxy.go

```diff
@@ -606,8 +611,17 @@ func (p *OAuthProxy) OAuthStart(rw http.ResponseWriter, req *http.Request) {
 		p.ErrorPage(rw, 500, "Internal Error", err.Error())
 		return
 	}
+
+	// clear cached endpoints, get fresh ones in case /.well-known/oauth-authorization-server changed its values
+	p.provider.ClearEndpointsCache()
+
 	redirectURI := p.GetRedirectURI(req.Host)
-	http.Redirect(rw, req, p.provider.GetLoginURL(redirectURI, fmt.Sprintf("%v:%v", nonce, redirect)), 302)
+	loginURL := p.provider.GetLoginURL()
+	if loginURL == nil {
+		p.ErrorPage(rw, 500, "Internal Error", fmt.Sprintf("could not get login endpoint"))
+		return
+	}
+	http.Redirect(rw, req, p.provider.GetLoginRedirectURL(*loginURL, redirectURI, fmt.Sprintf("%v:%v", nonce, redirect)), 302)
```

Comment:
Same here: let's put the cache clear and loginURL fetch in the `provider.GetLoginRedirectURL` function

Reply:
Unless I missed something about Go handling stuff, you can't have the `Get*()` functions inside the implementations of the providers (if you don't want to have duplicate implementation of it for every provider).

Reason (and this is why I was puzzled in https://github.com/openshift/oauth-proxy/pull/101#issuecomment-461081517):
`GetLoginRedirectURL()` is currently not implemented in `openshift.provider` package. So, if `GetLoginURL()` goes inside `GetLoginRedirectURL()`, this is what happens:
1. `p.provider` is `OpenShiftProvider`, which contains `ProviderData` struct. Therefore, from the outside, the MRO is `(OpenShiftProvider, ProviderData)`.
2. When you hit `GetLoginRedirectURL()`, the runtime looks in `OpenShiftProvider`, finds nothing, goes to `ProviderData` and finds the implementation there. It therefore continues inside the `(p *ProviderData) GetLoginRedirectURL()` method.
3. When go sees `p.GetLoginURL()` it then logically does `ProviderData.GetLoginURL()` - if you've just flushed cache, this will give you empty result (see the implementation of the method).

I tested this behavior, unless I did something incorrectly, I believe that is what happens.

I am more used to the Pythonic MRO, where
```python
class A:
  def getA(self):
    return "a"
  def getMe(self):
    return self.getA()

class B(A):
  def getA(self):
    return "b " + str(self.__class__)

o = B()
print(o.getMe())
```
As you'd expect, this second example prints `b <class '__main__.B'>`. As you can see in this case, Python holds the correct reference to the original object in the `self` variable, whereas we're only working with whatever we have inside, which is the reference to `ProviderData` and its implementations.

Reply:
Ugh, that's probably correct. This codebase... 😫